TASK 4.4

1.- Importing Libraries & Data

In [5]:
import pandas as pd
import numpy as np
import os

In [6]:
path = r'C:\Users\cmfag\OneDrive\Desktop\Data Analyst Specialist\Achievement 4\4.3_orders_products'

In [7]:
df_ords= pd.read_csv(os.path.join(path, 'orders.csv'), index_col = False)

In [8]:
df_prods= pd.read_csv(os.path.join(path, 'products.csv'), index_col = False)

In [9]:
df_deps= pd.read_csv(os.path.join(path, 'departments.csv'), index_col = False)

2.- Wrangling Data

In [11]:
# 2.1 identifiying not needed variables in the df_ords dataframe. 

df_ords.head()

# no needed varibale: eval_set

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
#removing varibale eval_set

df_ords.drop(columns = ['eval_set'])


order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [13]:
# 2.2 changing the data type of a variable that doesn´t need to be included as a numeric variable. 
  
df_ords.dtypes

# since 'order_id', 'user_id' and 'order_number' are numbers but need to be understand as a descrete variable, a change in the data type is needed.  


order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [14]:
#changing the data type-

df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')
df_ords['order_number'] = df_ords['order_number'].astype('str')


#varifing the result:

df_ords.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number               object
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

3.- Changing the name of a variable

In [16]:
#changing 'order_dow' to 'orders_day_of_the_week'

df_ords.rename (columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

#verifiying the output:

df_ords.head()


order_id user_id eval_set order_number  orders_day_of_week  \
0  2539329       1    prior            1                   2   
1  2398795       1    prior            2                   3   
2   473747       1    prior            3                   3   
3  2254736       1    prior            4                   4   
4   431534       1    prior            5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

4.- What the busiest hour is for placing orders?

In [18]:
# Calculating the frecuency of each order using the column 'order_hour_of_Day' 

hourly_orders = df_ords['order_hour_of_day'].value_counts()

print (hourly_orders)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64


In [19]:
# Encontrar la hora con más pedidos
busiest_hour = hourly_orders.idxmax()
frequency = hourly_orders.max()

print(busiest_hour, frequency)


10 288418


In [126]:
# printing the result

print(f"The busiest hour for placing orders is at {busiest_hour}:00, with {frequency} orders.")

The busiest hour for placing orders is at 10:00, with 288418 orders.


5 .- What is the meaning behind the value 4 in 'department_id'?

In [82]:
#creating the data dictionary

#first creating the df_departmen_id and then, the data dictionary

df_deps.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [98]:
df_deps_t=df_deps.T
new_header = df_deps_t.iloc[0]
df_deps_t_new = df_deps_t[1:]
df_deps_t_new.columns = new_header

In [100]:
df_deps_t_new.head

<bound method NDFrame.head of department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing>

In [120]:
#creating the data dictionary

data_dict_df_deps_t_new = df_deps_t_new.to_dict('index')

In [131]:
data_dict_df_deps_t_new

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [135]:
print (data_dict_df_deps_t_new.get ('4'))


# The meaning behind a value of 4 in the "department_id" column within the df_prods dataframe ist produce.

{'department': 'produce'}


6.- The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

Based on the data_dict_df_deps_t_new breakfast has the id: 14.

In [139]:
#creating the a  subset 

df_prods_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [145]:
#result: 

df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## 7.- Crearting a Diner Party Sunset - 

Find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [153]:
df_prods_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [155]:
df_prods_dinner_party.head

<bound method NDFrame.head of        product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id 

8.- It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [160]:
df_prods_dinner_party.shape

(7650, 5)

or

In [170]:
num_rows = df_prods_dinner_party.shape[0]

print(f"The new dinner party data frame has {num_rows} rows.")


The new dinner party data frame has 7650 rows.


9.- Information about user_id nr. 1

In [184]:
df_user1=df_ords.loc[df_ords['user_id'] == '1']
df_user1

order_id user_id eval_set order_number  orders_day_of_week  \
0   2539329       1    prior            1                   2   
1   2398795       1    prior            2                   3   
2    473747       1    prior            3                   3   
3   2254736       1    prior            4                   4   
4    431534       1    prior            5                   4   
5   3367565       1    prior            6                   2   
6    550135       1    prior            7                   1   
7   3108588       1    prior            8                   1   
8   2295261       1    prior            9                   1   
9   2550362       1    prior           10                   4   
10  1187899       1    train           11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

10.- User Behaviour

In [182]:
df_user1.describe()

orders_day_of_week  order_hour_of_day  days_since_prior_order
count           11.000000          11.000000               10.000000
mean             2.636364          10.090909               19.000000
std              1.286291           3.477198                9.030811
min              1.000000           7.000000                0.000000
25%              1.500000           7.500000               14.250000
50%              3.000000           8.000000               19.500000
75%              4.000000          13.000000               26.250000
max              4.000000          16.000000               30.000000

This user has a mixed shopping pattern, varying by days and times. Most of their orders happen early in the week, with an average day around Tuesday (2.64), and they usually order close to the start of the week. The majority of orders are made between morning and midday, with the average order time around 10 a.m. While there is some variation in the time of day (around 3.5 hours), most orders fall between 7 a.m. and 1 p.m. On average, the user waits 19 days between purchases, though this can range widely from 0 to 30 days. This shows that even though they tend to order at a similar time of day, the exact day and frequency are more flexible.

11.- Data Export

In [190]:
#export orders data frame
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [193]:
#export departments data frame
df_deps_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))